src/coreclr/tools/Common/TypeSystem/Ecma/EcmaModule.cs

```diff
@@ -282,60 +282,80 @@ public bool IsPlatformNeutral
             }
         }
 
+        private Dictionary<(string Name, string Namespace), EntityHandle> _nameLookupCache;
+
+        private Dictionary<(string Name, string Namespace), EntityHandle> CreateNameLookupCache()
+        {
+            // TODO: it's not particularly efficient to materialize strings just to hash them and hold
+            // onto them forever. We could instead hash the UTF-8 bytes and hold the TypeDefinitionHandle
+            // so we can obtain the bytes again when needed.
+            // E.g. see the scheme explored in the first commit of https://github.com/dotnet/runtime/pull/84285.
+
+            var result = new Dictionary<(string Name, string Namespace), EntityHandle>();
```

Comment:
If we care enough, this can be made more efficient by:
- Pre-sizing the dictionary - either conservatively using total counts in  `TypeDefinitionHandle` and `ExportedTypeHandle` tables, or precisely by doing a quick pass over these tables to see how many entries we are actually going to need.
- Using a custom struct with faster non-randomized GetHashCode algorithm (e.g. use the FNV hashcode implementation from Roslyn)

Reply:
Pre-sizing the dictionary can help a lot with GC throughput in my experience. That may matter in the case that we keep the code that does a GC before handing off to the object writer.

Reply:
If we go the custom hasher route, it's not that much extra work to do the thing I did in the first commit and also get rid of the dictionary in the first place.

We could also explore `FrozenDictionary` although it probably won't work for netstandard2.0 right now.

I don't have big concerns based on the perf numbers I collected. This is an improvement even when not reading MIBC (the perf numbers with Dictionary were collected without MIBC).